### ISSUE TITLE
[question] Is there a way to determine the commit SHA a conan package was built from?

### ISSUE THREAD

**Issue by RochaStratovan** [conan-io/conan#15882]
### What is your question?

Assuming a package is built from source code that is stored in a git repository (or any other SCM that has commit numbers, or commit SHAs, etc.) Is there a way that I can determine which SCM commit the package was built for?

I read the section "_[Capturing Git scm Information](https://docs.conan.io/2/examples/tools/scm/git/capture_scm/git_capture_scm.html)_" and saw how we could basically embed the scm information including exact SHA instance into the package for future creation (we need this short of solution, as source code access is restrictive for some projects, while the binaries are not).

I understand that the information is stored in a `conandata.yml` file. Is this available for our usage too? If so how would we access it? Or is there another way to obtain this information using `conan` CLI commands?

---

Prior to finding the SCM capture method, I was thinking of adding a custom option that was the commit SHA. Then I believe the information would be visible in the `conan list "*:*"` output.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @RochaStratovan 

Thanks for your question.

There is one alternative that you might find interesting, is using ``revision_mode = "scm"`` in which the recipe revision take the value directly from the git commit. The other alternative is what you mention, explicitly capturing the git commit in the ``conandata.yml``

> I understand that the information is stored in a conandata.yml file. Is this available for our usage too? If so how would we access it? Or is there another way to obtain this information using conan CLI commands?

yes, definitely. The recipes can read the value in ``self.conan_data`` and it is also available in the ``self.dependencies`` interface, so recipes, hooks, deployers, etc, can access the conan-data and the git commit of the dependencies stored there. At this moment it is not dumped in the ``json`` serialization of the graph, but it might be added if there is a use case for it.


> Prior to finding the SCM capture method, I was thinking of adding a custom option that was the commit SHA. Then I believe the information would be visible in the conan list "*:*" output.

Not a great idea, the options are to encode binary compatibility, but this is a "source" coordinate, not binary model, so it might not fit very well.

**Comment by RochaStratovan** [comment]
Hello @memsharded,

# Trying to use scm revision mode

I read about `revision_mode = "scm"` and I was fearful that it might not work. I saw the warning "_**it is critical that every revision represents an immutable source, including the recipe and the source code.**_"

My fear was for a scenario as outlined below (which is seemingly supported by my test):

```bash
scm_a <-- scm_b <--- scm_test
```

Where executable `scm_test` uses library `scm_b` which uses `scm_a`, we could get into a case where the same revision (SCM SHA) results in different results/packages. When `scm_a` updates to a new version that causes `scm_b` and `scm_test` to recompile and leads to new binaries.

I did the following test steps

1. **Create packages for the three elements and view their contents in the local cache.**

   * scm_a (1.0.0)
   * scm_b (2.0.0)
   * scm_test (3.0.0)

   <details><summary><fixed>conan list "scm*#*:*"</fixed> output</summary>

   ```bash
   $ conan list "scm*#*:*"
   Found 3 pkg/version recipes matching scm* in local cache
   Local Cache
     scm_a
       scm_a/1.0.0
         revisions
           ad5740074fe4aba6e0e2786ba0890b38a98eb99f (2024-03-18 23:09:02 UTC)
             packages
               9bdee485ef71c14ac5f8a657202632bdb8b4482b
                 info
                   settings
                     arch: x86_64
                     build_type: Release
                     compiler: msvc
                     compiler.cppstd: 14
                     compiler.runtime: dynamic
                     compiler.runtime_type: Release
                     compiler.version: 193
                     os: Windows
                   options
                     shared: False
     scm_b
       scm_b/2.0.0
         revisions
           30e0b948cb6adf93dce4b2be73627c7e84d6aadd (2024-03-18 23:09:21 UTC)
             packages
               c276027dbabff7baa3a1e4e307a706f80cbc079f
                 info
                   settings
                     arch: x86_64
                     build_type: Release
                     compiler: msvc
                     compiler.cppstd: 14
                     compiler.runtime: dynamic
                     compiler.runtime_type: Release
                     compiler.version: 193
                     os: Windows
                   options
                     shared: False
                   requires
                     scm_a/1.0.Z
     scm_test
       scm_test/3.0.0
         revisions
           9445568232378f6e6e8461e14f645d287a7e9bcc (2024-03-18 23:10:45 UTC)
             packages
               08b67a7c85f84d6174f61d3e7155591b243df0c3
                 info
                   settings
                     arch: x86_64
                     build_type: Release
                     compiler: msvc
                     compiler.cppstd: 14
                     compiler.runtime: dynamic
                     compiler.runtime_type: Release
                     compiler.version: 193
                     os: Windows
                   requires
                     scm_a/1.0.0#ad5740074fe4aba6e0e2786ba0890b38a98eb99f:9bdee485ef71c14ac5f8a657202632bdb8b4482b
                     scm_b/2.0.0#30e0b948cb6adf93dce4b2be73627c7e84d6aadd:c276027dbabff7baa3a1e4e307a706f80cbc079f
   ```

   </details>


1. Make a change to the lowest level library **scm_a**. Which updates the SHA and triggers a new revision. And then simulate downstream compiles by issuing `conan create` for `scm_b` (because scm_a changed) and `scm_test` because scm_b will now have changed.

   <details><summary><fixed>conan list "scm*#*:*"</fixed> output</summary>

   ```bash
   h$ conan list "scm*#*:*"
   Found 4 pkg/version recipes matching scm* in local cache
   Local Cache
     scm_a
       scm_a/1.0.0
         revisions
           ad5740074fe4aba6e0e2786ba0890b38a98eb99f (2024-03-18 23:09:02 UTC)
             packages
               9bdee485ef71c14ac5f8a657202632bdb8b4482b
                 info
                   settings
                     arch: x86_64
                     build_type: Release
                     compiler: msvc
                     compiler.cppstd: 14
                     compiler.runtime: dynamic
                     compiler.runtime_type: Release
                     compiler.version: 193
                     os: Windows
                   options
                     shared: False
       scm_a/1.0.1
         revisions
           8e6d883b0b3346409b0b52cdb35a1211c4e3f6f6 (2024-03-18 23:13:20 UTC)
             packages
               9bdee485ef71c14ac5f8a657202632bdb8b4482b
                 info
                   settings
                     arch: x86_64
                     build_type: Release
                     compiler: msvc
                     compiler.cppstd: 14
                     compiler.runtime: dynamic
                     compiler.runtime_type: Release
                     compiler.version: 193
                     os: Windows
                   options
                     shared: False
     scm_b
       scm_b/2.0.0
         revisions
           30e0b948cb6adf93dce4b2be73627c7e84d6aadd (2024-03-18 23:14:40 UTC)
             packages
               c276027dbabff7baa3a1e4e307a706f80cbc079f
                 info
                   settings
                     arch: x86_64
                     build_type: Release
                     compiler: msvc
                     compiler.cppstd: 14
                     compiler.runtime: dynamic
                     compiler.runtime_type: Release
                     compiler.version: 193
                     os: Windows
                   options
                     shared: False
                   requires
                     scm_a/1.0.Z
     scm_test
       scm_test/3.0.0
         revisions
           9445568232378f6e6e8461e14f645d287a7e9bcc (2024-03-18 23:16:40 UTC)
             packages
               08b67a7c85f84d6174f61d3e7155591b243df0c3
                 info
                   settings
                     arch: x86_64
                     build_type: Release
                     compiler: msvc
                     compiler.cppstd: 14
                     compiler.runtime: dynamic
                     compiler.runtime_type: Release
                     compiler.version: 193
                     os: Windows
                   requires
                     scm_a/1.0.0#ad5740074fe4aba6e0e2786ba0890b38a98eb99f:9bdee485ef71c14ac5f8a657202632bdb8b4482b
                     scm_b/2.0.0#30e0b948cb6adf93dce4b2be73627c7e84d6aadd:c276027dbabff7baa3a1e4e307a706f80cbc079f
               6b580b972a65c6aedccd889ad4fa5bdf6a9d631c
                 info
                   settings
                     arch: x86_64
                     build_type: Release
                     compiler: msvc
                     compiler.cppstd: 14
                     compiler.runtime: dynamic
                     compiler.runtime_type: Release
                     compiler.version: 193
                     os: Windows
                   requires
                     scm_a/1.0.1#8e6d883b0b3346409b0b52cdb35a1211c4e3f6f6:9bdee485ef71c14ac5f8a657202632bdb8b4482b
                     scm_b/2.0.0#30e0b948cb6adf93dce4b2be73627c7e84d6aadd:c276027dbabff7baa3a1e4e307a706f80cbc079f
   ```

   </details>


I think these results are showing (as I expected) that the packages are unique. However, maybe I am misinterpreting what is occurring and you can teach me more.

What I see is:

1. `scm_b` has one revision but they look to be different.

   |           | scm_b #<z>1                                 | scm_b #<z>2                                 |
   |-----------|------------------------------------------|------------------------------------------|
   | revision  | 30e0b948cb6adf93dce4b2be73627c7e84d6aadd | 30e0b948cb6adf93dce4b2be73627c7e84d6aadd |
   | **timestamp** | 2024-03-18 23:09:21 UTC                  | 2024-03-18 23:14:40 UTC                  |
   | package   | c276027dbabff7baa3a1e4e307a706f80cbc079f | c276027dbabff7baa3a1e4e307a706f80cbc079f |

   This is alarming to me because the timestamp changed. Although I'm speculating this is a case of, the library itself hasn't changed, but we did issue create so the same library was updated into the cache at this new time.

1. `scm_test` has two packages

   |           | scm_test #<z>1                              | scm_test #<z>2                              |
   |-----------|------------------------------------------|------------------------------------------|
   | revision  | 9445568232378f6e6e8461e14f645d287a7e9bcc | 9445568232378f6e6e8461e14f645d287a7e9bcc |
   | **timestamp** | 2024-03-18 23:10:45 UTC                  | 2024-03-18 23:16:40 UTC                                         |
   | **package**   | 08b67a7c85f84d6174f61d3e7155591b243df0c3<br> | 08b67a7c85f84d6174f61d3e7155591b243df0c3<br>6b580b972a65c6aedccd889ad4fa5bdf6a9d631c |

   I find this alarming because the revision view "`conan list "scm*#*`" only shows one revision for version. Which makes sense because the SHA doesn't change. But the detailed view shows that I have two packages for `scm_test` which makes sense as #<z>1 uses scm_a/1.0.0 and #<z>2 uses scm_a/1.0.1. 
   
   Previously I was applying the mental rule "we'll use the last revision for a selected version". How would that apply here? There is one revision for scm_test but it has two packages.


I was able to cause the same sort of scenario but with the `scm_b` library to have two packages. I just created an API change version for `scm_a/1.1.0`.


# using the conandata.yml

How can I access the contents of the `conandata.yml` without running `conan install`, `conan build`, etc. For example, if I've downloaded the package file from a remote, and I want to learn what SCM the package is for, how would I do that? The package will have been imported into my local cache via a `conan install` or `conan build` for a dependent project or because I did a `conan download`.

You're explaining methods that would allow my conan recipe file to access the information. I want to know how to access it as an external user typing instructions into my CLI.

# Don't recommend adding SHA as a commit option

You say this isn't a good idea because options are for binary compatibility information. My naive response is, so what? How does this hurt anything? What is the actual negative impact of doing something like this? From my naive perspective it sounds like you're saying this isn't how conan intended for this to be used. To which my response is, okay, but will it hurt anything? Break anything? Cause any other problems? Why would I not do this to make it easy to learn the SHA for a package?

**Comment by memsharded** [comment]
> scm_b has one revision but they look to be different.

This is expected and looks correct. 
The important bit is that the ``recipe-revision`` and the ``package-id`` are still the same. As they are static library, a revision change in a dependency, which is a static library shouldn't result in re-building ``scm_b`` at all.

The reason that you are seeing a revision timestamp change might be because you are doing a ``conan export`` or ``conan create`` again, which exports, and then assigns a new updated timestamp to the same thing. If you did a ``conan install --requires=scm_b/... --build=missing``, there would be no timestamp update at all, it will be a no-op.

You can see the "independence" of the version of ``scm_a`` in:
```
 requires
      scm_a/1.0.Z
```

> scm_test has two packages

This looks correct too.
The recipe revision is the same in ``scm_test`` because the source code is exactly the same.

But then it gets 2 different binaries with 2 different ``package_ids`` because there are 2 variants of the binary, built and linked against different dependencies binaries. You can see it in:
```
requires
        scm_a/1.0.0#ad5740074fe4aba6e0e2786ba0890b38a98eb99f:9bdee485ef71c14ac5f8a657202632bdb8b4482b
        scm_b/2.0.0#30e0b948cb6adf93dce4b2be73627c7e84d6aadd:c276027dbabff7baa3a1e4e307a706f80cbc079f
```

All of this assumes that the change in ``scm_a`` is not affecting the API or external ABI at all, mostly the headers shouldn't be modified and changes are only in private implementation files. If the public headers change, it is likely that a version bump is more recommended to represent that change in the public API/ABI

**Comment by memsharded** [comment]
> You're explaining methods that would allow my conan recipe file to access the information. I want to know how to access it as an external user typing instructions into my CLI.

Not sure what is the intention here.
Any consumer recipe can access ``self.dependencies["mydep"].conan_data`` of its dependencies, for example to write or generate some file or the like. Is this what you are looking for?

**Comment by RochaStratovan** [comment]
Hello @memsharded,

# Not sure what the intention is

> Not sure what is the intention here.
> Any consumer recipe can access self.dependencies["mydep"].conan_data of its dependencies, for example to write or generate some file or the like. Is this what you are looking for?

First I need to say we don't currently use conan. For the following, when I say cache, I'm referring to my company's current mechanism. However, conan would replace this and is fairly similar to our existing home grown solution.

I frequently need to help a developer debug. They often insist that they have a version that includes the fix. And looking at the commit SHA we can determine if this is correct or not. Many of our developers work remotely so they update their cache from mothership once or twice a week and work from there.

I need a simple, quick, reliable way to be able to view the developer's cache (in the future that would be `conan list` information) and determine which commit SHA their package comes from.

Given that a developer has the package `foo` installed to their cache because they did one of the following

1. `conan download foo` to directly install `foo` to their cache

1. `conan install blarg` to install dependencies for project `blarg` and it happens to use `foo`.

1. `conan build blarg` (same as the previous case).

How would I or a developer be able to determine what commit SHA the package `foo` was built from? At this point in time what we have is a conan cache with `foo` and maybe the project `blarg`. Is there a conan CLI command to allow me to determine `foo`'s commit SHA?

In re-reading your answer, I _**think**_ you're saying Conan does not have this ability via it's CLI. However, we could use Conan's programmatic interface to achieve this. For example `blarg`'s `conanfile.py` could create a metadata file during `conan install` or `conan build` where the recipe logic uses `self.dependencies["foo"].conan_data` to place the SCM SHA info into a metadata file. Or we could utilize the ability to add a new conan command that would do this.

**TL;DR;** I think you're saying Conan commands do not do what we want, but we can add our own conan command or modify conan recipes to do this.



# More questions from my example

Can you comment on the scenario when `scm_a` had an API change which then causes two packages for `scm_b`?

<details><summary><tt>conan list "scm*#*</tt></summary>

```bash
$ conan list "scm*#*"
Found 5 pkg/version recipes matching scm* in local cache
Local Cache
  scm_a
    scm_a/1.0.0
      revisions
        ad5740074fe4aba6e0e2786ba0890b38a98eb99f (2024-03-18 23:09:02 UTC)
    scm_a/1.0.1
      revisions
        8e6d883b0b3346409b0b52cdb35a1211c4e3f6f6 (2024-03-18 23:13:20 UTC)
    scm_a/1.1.0
      revisions
        8e6d883b0b3346409b0b52cdb35a1211c4e3f6f6 (2024-03-18 23:45:22 UTC)
  scm_b
    scm_b/2.0.0
      revisions
        30e0b948cb6adf93dce4b2be73627c7e84d6aadd (2024-03-18 23:45:33 UTC)
  scm_test
    scm_test/3.0.0
      revisions
        9445568232378f6e6e8461e14f645d287a7e9bcc (2024-03-18 23:45:41 UTC)
```

</details>

<details><summary><tt>conan list "scm*#*:*"</tt></summary>

```bash
$ conan list "scm*#*:*"
Found 5 pkg/version recipes matching scm* in local cache
Local Cache
  scm_a
    scm_a/1.0.0
      revisions
        ad5740074fe4aba6e0e2786ba0890b38a98eb99f (2024-03-18 23:09:02 UTC)
          packages
            9bdee485ef71c14ac5f8a657202632bdb8b4482b
              info
                settings
                  arch: x86_64
                  build_type: Release
                  compiler: msvc
                  compiler.cppstd: 14
                  compiler.runtime: dynamic
                  compiler.runtime_type: Release
                  compiler.version: 193
                  os: Windows
                options
                  shared: False
    scm_a/1.0.1
      revisions
        8e6d883b0b3346409b0b52cdb35a1211c4e3f6f6 (2024-03-18 23:13:20 UTC)
          packages
            9bdee485ef71c14ac5f8a657202632bdb8b4482b
              info
                settings
                  arch: x86_64
                  build_type: Release
                  compiler: msvc
                  compiler.cppstd: 14
                  compiler.runtime: dynamic
                  compiler.runtime_type: Release
                  compiler.version: 193
                  os: Windows
                options
                  shared: False
    scm_a/1.1.0
      revisions
        8e6d883b0b3346409b0b52cdb35a1211c4e3f6f6 (2024-03-18 23:45:22 UTC)
          packages
            9bdee485ef71c14ac5f8a657202632bdb8b4482b
              info
                settings
                  arch: x86_64
                  build_type: Release
                  compiler: msvc
                  compiler.cppstd: 14
                  compiler.runtime: dynamic
                  compiler.runtime_type: Release
                  compiler.version: 193
                  os: Windows
                options
                  shared: False
  scm_b
    scm_b/2.0.0
      revisions
        30e0b948cb6adf93dce4b2be73627c7e84d6aadd (2024-03-18 23:45:33 UTC)
          packages
            356cd90cb388afb2b4f3de61cb60af867a5c641b
              info
                settings
                  arch: x86_64
                  build_type: Release
                  compiler: msvc
                  compiler.cppstd: 14
                  compiler.runtime: dynamic
                  compiler.runtime_type: Release
                  compiler.version: 193
                  os: Windows
                options
                  shared: False
                requires
                  scm_a/1.1.Z
            c276027dbabff7baa3a1e4e307a706f80cbc079f
              info
                settings
                  arch: x86_64
                  build_type: Release
                  compiler: msvc
                  compiler.cppstd: 14
                  compiler.runtime: dynamic
                  compiler.runtime_type: Release
                  compiler.version: 193
                  os: Windows
                options
                  shared: False
                requires
                  scm_a/1.0.Z
  scm_test
    scm_test/3.0.0
      revisions
        9445568232378f6e6e8461e14f645d287a7e9bcc (2024-03-18 23:45:41 UTC)
          packages
            08b67a7c85f84d6174f61d3e7155591b243df0c3
              info
                settings
                  arch: x86_64
                  build_type: Release
                  compiler: msvc
                  compiler.cppstd: 14
                  compiler.runtime: dynamic
                  compiler.runtime_type: Release
                  compiler.version: 193
                  os: Windows
                requires
                  scm_a/1.0.0#ad5740074fe4aba6e0e2786ba0890b38a98eb99f:9bdee485ef71c14ac5f8a657202632bdb8b4482b
                  scm_b/2.0.0#30e0b948cb6adf93dce4b2be73627c7e84d6aadd:c276027dbabff7baa3a1e4e307a706f80cbc079f
            6b580b972a65c6aedccd889ad4fa5bdf6a9d631c
              info
                settings
                  arch: x86_64
                  build_type: Release
                  compiler: msvc
                  compiler.cppstd: 14
                  compiler.runtime: dynamic
                  compiler.runtime_type: Release
                  compiler.version: 193
                  os: Windows
                requires
                  scm_a/1.0.1#8e6d883b0b3346409b0b52cdb35a1211c4e3f6f6:9bdee485ef71c14ac5f8a657202632bdb8b4482b
                  scm_b/2.0.0#30e0b948cb6adf93dce4b2be73627c7e84d6aadd:c276027dbabff7baa3a1e4e307a706f80cbc079f
            b39ad79d1e72a0eecfdca8e1fa1cfa77276210ee
              info
                settings
                  arch: x86_64
                  build_type: Release
                  compiler: msvc
                  compiler.cppstd: 14
                  compiler.runtime: dynamic
                  compiler.runtime_type: Release
                  compiler.version: 193
                  os: Windows
                requires
                  scm_a/1.1.0#8e6d883b0b3346409b0b52cdb35a1211c4e3f6f6:9bdee485ef71c14ac5f8a657202632bdb8b4482b
                  scm_b/2.0.0#30e0b948cb6adf93dce4b2be73627c7e84d6aadd:356cd90cb388afb2b4f3de61cb60af867a5c641b
```

</details>

This feels like it breaks the immutable rule for a revision.

In this case `scm_b` has one revision with two packages. How would I tell from the output which one should be used? How can I tell which is most recent? They both share the same timestamp.

I don't think order is the answer as the output appears to be sorted based on the package id value.

The long and short of it for me is that _scm mode_ feels unreliable/unpredictable. The documentation, commands and demos make sense to me down to the revision level, but then beyond that it gets muddy to me.

**Comment by memsharded** [comment]
> I need a simple, quick, reliable way to be able to view the developer's cache (in the future that would be conan list information) and determine which commit SHA their package comes from.

It is important to highlight that the Conan cache might actually have multiple revisions for the package. So it is unlikely that a ``conan list`` that list those revisions (or only the latest) will be able to give a good information about this. The information needs to be explicit:
- This specific ``mypkg/version#specific_revision``, irrespective of whether the revision is the scm git-commit one or the contents hash
- When testing things as a part of a whole for reproduction, then the full ``lockfile`` might be necessary. This lockfile will also contain the specific ``mypkg/version#specific_revision`` that the developer is testing.

In that regard, the Conan recipe-revision is similar to a commit, there will be more than one commit, and the developer can checkout different branches, so listing all branches in the developer machine is not enough, it is necessary to know exactly the recipe-revision at hand.

> How would I or a developer be able to determine what commit SHA the package foo was built from? At this point in time what we have is a conan cache with foo and maybe the project blarg. Is there a conan CLI command to allow me to determine foo's commit SHA?

If the ``recipe-revision`` is the commit-hash, then the response is direct, nothing to check there.
If not, there could be different approaches:
```
conan cache path mypkg/version#recipe_revision
```
will give the folder that contains the ``conandata.yml``, so reading it and obtaining the git commit is straightforward.
We are already considering making this more built-in, but it sounds that adding it as metadata would probably be too much, making it part of the ``conandata`` is also better, in the sense that this also gives the recipe the capability of building from source from that exact commit too, it is part of the necessary ``source()`` information.

**Comment by memsharded** [comment]
For the other question:

> This feels like it breaks the immutable rule for a revision.

> In this case scm_b has one revision with two packages. How would I tell from the output which one should be used? How can I tell which is most recent? They both share the same timestamp.

The immutable rule works at the recipe-revision level in this case. The sources or recipe for ``scm_b`` didn't change, so the revision is the same, the recipe is still immutable.

The timestamp is not relevant here, because there is not a relationship of time precedence between 2 different ``package_ids``. They represent different configurations, not evolution over time. There is no time-precendence of a new configuration over an older one, for example you can have 2 binaries for the same library, one for windows and the other for linux. The fact that the compilation of the windows binary finished earlier than the linux one in CI is irrelevant, because package binaries are not resolved by timestamps, but by configuration.

With dependencies versions happens the same. The timestamp does not resolve anything (if you are curious, you can still list it with ``conan list pkg/version:*#*``), what it resolves is the versioning approach. If ``scm_a/1.1.0`` is considered always newer than ``scm_a/1.0.0``, any version range that allows both will always resolve to the newest ``scm_a/1.1.0`` and that will naturally and automatically result in the new ``package_id`` that was computed against that version of the dependency, that lists:
```
requires
     scm_a/1.1.Z
```
If on the other hand, you pass to the graph a lockfile that forces to stick with the ``scm_a/1.0.0`` version, then ``scm_b`` will naturally result in the other ``package_id``, in the same way that if you pass ``-s os=Windows`` or ``-s os=Linux`` in the profile/cli.

> The long and short of it for me is that scm mode feels unreliable/unpredictable. The documentation, commands and demos make sense to me down to the revision level, but then beyond that it gets muddy to me.

Not really. It seems we are mixing here concepts of "source" origin and package binaries with ``package_id``. At the source level the ``recipe-revision`` achieves "source immutability", but that is an orthogonal aspect to having different package binaries built with exactly the same source input.

**Comment by RochaStratovan** [comment]
Thank you for your continued patience with this.

# Is this what you mean?

I am getting lost, can verify what I think you are saying?

Consider the following:

```bash
foo <-- blarg
```

1. The project `blarg` uses package `foo`.

1. We do _**not**_ use SCM mode for revisions.

1. We _**do**_ use [Capturing Git scm information](https://docs.conan.io/2/examples/tools/scm/git/capture_scm/git_capture_scm.html) to update the `conandata.yml` file.

1. Developer is working on `blarg`

1. Developer does `conan install blarg` to update their local cache and prepare for their local development.

1. Developer now develops using appropriate flags to pull in the Conan generated CMake configuration files, etc.

At this point we have a problem and need to debug. We know the project `blarg` uses the package `foo` and we suspect there's an issue with `foo`. We're trying to get information on `foo`.

We want to know what commit the `foo` package that we're working with correspond to. The commit SHA allows us to go to the corresponding git repository and determine if it is the latest commit. If not the latest commit, are there changes after this that might mention or resolve the issue being observed. Or maybe a recent commit made changes in the area that I'm seeing the issue, etc.

You mention using the command

```bash
conan cache path foo/<version>#<revision>
```

to obtain the path to the `conandata.yml` file.

_**Thank you.**_ I've done this for my local demo and observed the `conandata.yml` file contents!

<br>

_**Follow up question:**_ In my example I had one instance of the `foo` package installed so it was easy to identify the `foo/<version>#<revision>` that I needed to use.

You mentioned there could be multiple versions of the `foo` package in the conan cache. Given the scenario above, how would the developer know which version of the `foo` packge is being used? Is this a case where I would use say `conan graph build-order .` which shows the requirements information which includes `<version>` and `<revision>`?

_Summary of steps (I think):_

1. `conan graph build-order .` to get the `foo/<version>#<revision>` information

1. `conan cache path foo/<version>#<revision>` to get the path to the `conandata.yml` file

1. Use `<foo-cach-path>/conandata.yml` to get the stored SHA.

<br>

# SCM Mode Scenario Follow up

Thank you. It's still not 100% clear to me, but it starting to become more clear. And I think I play more with this to see what I learn.

**Comment by memsharded** [comment]
> Thank you for your continued patience with this.

No problem, happy to help 🙂 

> Thank you. I've done this for my local demo and observed the conandata.yml file contents!

Excellent, thanks for the feedback.

If you want to know how we are thinking, we are considering for the future workspace functionality to have a command that using ``conan open foo/version#revision`` will automatically do that, read the ``conandata.yml`` inside the recipe, then do a ``git clone + checkout`` automatically in user folder to allow local development of such exact revision. As it will do a checkout, if the dev wants to checkout and try a different commit, always possible.

> You mentioned there could be multiple versions of the foo package in the conan cache. Given the scenario above, how would the developer know which version of the foo packge is being used? Is this a case where I would use say conan graph build-order . which shows the requirements information which includes <version> and <revision>?

Might be even simpler:
- When doing a ``conan install blargh``, that command output will directly list the exact revision ``foo/version#revision`` too.
- It is also typical that when a developer wants to achieve more stability, will do ``conan install blargh --lockfile-out=conan.lock``. This ``conan.lock`` will also contain the exact revision ``foo/version#revision``

So not need to do an extra ``conan graph build-order``.

**Comment by ryan-thurber** [comment]
@memsharded I'm struggling with a related issue - is it possible to consume the git hash information of a dependency via the `self.dependencies["dep"].conan_data` in the `generate()` function when the dependent package is in editable mode?

For context, I am integrating Conan into several existing projects. I have a conan package for library `A`, and a conan package for executable `B` which depends on `A`. Both use the git scm information for source and export. I'd like to have the git hash from the resolved dependency `A` passed to the cmake toolchain at generate time, but it seems that it is not possible to do so when `A` is in editable mode, since exported data is not used, and `self.dependencies["dep"].conan_data` resolves to `None`.

Sorry if the question is unclear at all, I'm still learning Conan as I go 😁

**Comment by memsharded** [comment]
Hi @ryan-thurber 

> I'm struggling with a related issue - is it possible to consume the git hash information of a dependency via the self.dependencies["dep"].conan_data in the generate() function when the dependent package is in editable mode?

I am afraid not. When in ``editable`` mode, the conandata hasn't been filled yet with the commit information, that happens at ``export`` time.
The ``conan_data`` will not be ``None`` if there is a local ``conandata.yml`` file besides the ``conanfile.py``, and it will be automatically loaded and used. But the commit SHA will not be there, because the idea is that the commit hasn't been done yet, because you are in editable mode, and there are local changes that haven't been committed yet.
I think if you are using that information in the consumer, you might want to default to an arbitrary "under_edition" commit to represent this scenario.

What are you trying to achieve? Because one thing is that packages store their commit in their conandata, and something different is that consumer packages need to access that information. Why would a consumer need its dependencies commits for building itself?

**Comment by ryan-thurber** [comment]
@memsharded Thanks for the response!

> What are you trying to achieve? Because one thing is that packages store their commit in their conandata, and something different is that consumer packages need to access that information. Why would a consumer need its dependencies commits for building itself?

Our final executable links in a lot of our custom libraries, so one of the things we do for traceability purposes is compile in the git hashes of the the libraries that have been used. These hashes are then printed out in the the executable logs. This is obviously less than ideal, which is part of why I'm integrating Conan as a solution for dependency management, but I'd like to maintain it for a while since we have some tools that parse this information.

> But the commit SHA will not be there, because the idea is that the commit hasn't been done yet, because you are in editable mode, and there are local changes that haven't been committed yet.

This makes sense. I suppose this is a limitation I'll just have to accept, because it's also a gap in our current system, which relies on the repositories being checked out in a tree structure and does not check if the repo is clean before grabbing the git hash.

At this point I've been setting it up so that our developers will work with editable mode most of the time, but I may switch to the [Local Package Developer Flow](https://docs.conan.io/2/tutorial/developing_packages/local_package_development_flow.html) so that the package `A` is exported before compiling `B`. My main concern is the cost of caching every package change (both in i/o and storage), since our developers will be constantly recompiling these libraries.

**Comment by memsharded** [comment]
> This makes sense. I suppose this is a limitation I'll just have to accept, because it's also a gap in our current system, which relies on the repositories being checked out in a tree structure and does not check if the repo is clean before grabbing the git hash.

I guess that it might be possible to implement some smarter fallback.
When packages are in editable, you can still probably access their folder via ``self.dependencies["mydep"].package_folder`` or ``.recipe_folder``. Obtaining from there the git commit (even if dirty) is doable. So you could have a fallback that does that if ``conan_data`` doesn't contain the ``scm`` information. 
And then put it in a ``python_requires`` if you need to reuse that logic across many package recipes.

Feel free to open new ticket for further discussion about this if you need help, it will be easier to track that here in a closed ticket. Thanks!
